go-git/go-git
# plumbing: transport/ssh, auto-populate HostKeyAlgorithms. Fixes #411

This PR adjusts the transport/ssh logic in `command.connect()`, populating `ssh.ClientConfig.HostKeyAlgorithms` automatically now. The algorithms are chosen based on the known host keys for the target host, as obtained from the known_hosts file.

Because HostKeyAlgorithms vary by target host (based on the known_hosts entries for that specific hostname), some related logic for setting HostKeyCallback has been moved out of the various AuthMethod implementations in this PR. This was necessary because the old HostKeyCallbackHelper is not host-specific. Since known_hosts handling isn't really tied to AuthMethod anyway, it seems reasonable to separate these. Previously-exported types/methods remain in place for backwards compat, but some of them are now unused.

In order to look-up the algorithms from the known_hosts file, external module [github.com/skeema/knownhosts](https://github.com/skeema/knownhosts) is used. This package is just a thin wrapper around [golang.org/x/crypto/ssh/knownhosts](https://golang.org/x/crypto/ssh/knownhosts), adding an extra mechanism to query the known_hosts keys, implemented in a way which avoids duplication of any [golang.org/x/crypto/ssh/knownhosts](https://golang.org/x/crypto/ssh/knownhosts) logic.

Although [github.com/skeema/knownhosts](https://github.com/skeema/knownhosts) is a brand new package, the underlying code is already battle-tested: it has been extracted from the codebase of Skeema Premium CLI, which is used and trusted by a number of well-known companies as their mysql schema management solution. (For backstory, we recently added SSH tunnel support to Skeema, and were bit by https://github.com/golang/go/issues/29286 which is the same root cause as #411. We decided to open-source [github.com/skeema/knownhosts](https://github.com/skeema/knownhosts) as a generic solution to that problem.)

## Testing
Issue #411 can only be reproduced via end-to-end / integration testing, since it requires actually launching an SSH connection, in order to see the key mismatch error triggered from https://github.com/golang/go/issues/29286 as the root cause. In theory it is doable with Docker (installing `openssh-server` in a container -- that's what we do in Skeema for e2e SSH tunnel testing) but I'm afraid I'm not familiar enough with go-git's testing suites to attempt that here.

That said, this PR can be manually tested as follows:
1. Edit your ~/.ssh/known_hosts. For github.com, remove the ecdsa-sha2-nistp256 entry, but ensure at least one other entry (ssh-rsa or ssh-ed25519) is still present. This is to duplicate the conditions triggering #411.
2. On go-git master, run `go build` on _examples/clone/auth/ssh and use the resulting binary to attempt to clone any github repo. You will get an error "error: ssh: handshake failed: knownhosts: key mismatch"
3. Switch to the branch of this PR, run `go build` again on _examples/clone/auth/ssh and try the clone attempt again. This time it will succeed without error.
